lib/libsmb2.c

```diff
@@ -2791,6 +2796,145 @@ smb2_oplock_break_notify(struct smb2_context *smb2, int status, void *command_da
+    uint32_t next_entry_offset)
+{
+        uint32_t name_len;
+        
+        smb2_get_uint32(vec, next_entry_offset+4, &fnc->action);
+        smb2_get_uint32(vec, next_entry_offset+8, &name_len);
+        fnc->name = smb2_utf16_to_utf8((uint16_t *)&vec->buf[next_entry_offset+12], name_len / 2);
+
+        smb2_get_uint32(vec, next_entry_offset, &next_entry_offset);
+        if (next_entry_offset != 0) {
+                struct smb2_file_notify_change_information *next_fnc = calloc(1, sizeof(struct smb2_file_notify_change_information));
+                fnc->next = next_fnc;
+                smb2_decode_filenotifychangeinformation(smb2, next_fnc, vec, next_entry_offset);
+        }
+        return 0;
+}
+
+void
+free_smb2_file_notify_change_information(struct smb2_context *smb2, struct smb2_file_notify_change_information *fnc)
+{
+        if (fnc->next) {
+                free_smb2_file_notify_change_information(smb2, fnc->next);
+        }
+        free(discard_const(fnc->name));
+        free(fnc);
+}
+
+struct notify_change_cb_data {
+        smb2_command_cb cb;
+        void *cb_data;
+        // smb2fh file handle of the directory to get notified
+        struct smb2fh *fh;
+        // filter of SMB2_CHANGE_NOTIFY_FILE_NOTIFY_CHANGE_* flags
+        uint16_t filter;
+        // flags such as SMB2_CHANGE_NOTIFY_WATCH_TREE
+        uint32_t flags;      
+        // do a new notify_change request after each response if 1
+        uint32_t loop;
+        uint32_t status;
+};
+
+static void
+notify_change_cb(struct smb2_context *smb2, int status,
+          void *command_data _U_, void *private_data)
+{
+        struct notify_change_cb_data *notify_change_data = private_data;
+
+        struct smb2_change_notify_reply *rep = command_data;
+        struct smb2_iovec vec;
+        struct smb2_file_notify_change_information *fnc = calloc(1, sizeof(struct smb2_file_notify_change_information));
+
+        if (status) {
+                printf("notify_change_cb failed (%s) %s\n",
+                        strerror(-status), smb2_get_error(smb2));
+        }
+
+        vec.buf = rep->output;
+        vec.len = rep->output_buffer_length;
+
+        if (smb2_decode_filenotifychangeinformation(smb2, fnc, &vec, 0)) {
+                printf("Failed to decode file notify change information\n");
+        }
+
+        if (notify_change_data->cb) {
+                notify_change_data->cb(
+                        smb2,
+                        -nterror_to_errno(notify_change_data->status),
+                        fnc,
+                        notify_change_data->cb_data
+                );
+        }
+        if (notify_change_data->loop) {
+                smb2_notify_change_filehandle_async(smb2, notify_change_data->fh, notify_change_data->flags, notify_change_data->filter, 
+                        notify_change_data->loop, notify_change_data->cb, notify_change_data->cb_data);
+        } else {
+                smb2_close(smb2, notify_change_data->fh);
```

Comment:
I believe calling `smb2_notify_change_filehandle_async` should not result in closed file handle, whule it is reasonable to be closed if `smb2_notify_change_async` is called.
Can we put a flag in command data for this case?

Reply:
smb2_close is only called if notify_change_data->loop is set to false, which means it is supposed to be a one-off operation, and in that case it should be normal to close the file handle, as it is opened by smb2_notify_change_async

Reply:
I can call `smb2_notify_change_filehandle_async` with loop = false with a filehandle created by myself and this method should not close the handle in this case IMO.

Reply:
I agree with amosavian here. The function should not close the handle but I understand why it does.
The smb2_notify_change_async() function is just a small wrapper that takes a path and converts it to a filehandle and then calls the actual notify function, which takes a file handle.

Probably get rid of smb2_notify_change_async() and thus also the reason for the loop variable or closing the handle in the callback.
And then rename smb2_notify_change_filehandle_async() to smb2_notify_change_async().

Reply:
Another point I think that should be addressed is to check if the filehandle is still valid and open before calling `smb2_notify_change_filehandle_async` again.
@tryggvil @sahlberg